cirq-core/cirq/circuits/circuit.py

```diff
@@ -1534,10 +1538,8 @@ def factorize(self: CIRCUIT_TYPE) -> Iterable[CIRCUIT_TYPE]:
         )
 
     def _control_keys_(self) -> FrozenSet[value.MeasurementKey]:
-        all_keys: FrozenSet[value.MeasurementKey] = frozenset()
-        for op in self.all_operations():
-            all_keys = all_keys.union(protocols.control_keys(op))
-        return all_keys.difference(protocols.measurement_key_objs(self))
+        controls = frozenset(k for op in self.all_operations() for k in protocols.control_keys(op))
+        return controls - protocols.measurement_key_objs(self)
```

Comment:
To confirm: a circuit which contains control and measurement of key "A" excludes "A" from what it returns here because "A" is _internal_, correct? (i.e. it's measured and used entirely within the circuit)

Reply:
Yes, the protocol returns all keys that must be defined externally before this element can be used.